<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Introduction to DESeq2 &mdash; Duke NGS Course (Summer 2015) 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-3.2.0/css/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-3.2.0/css/bootstrap-theme.min.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="_static/bootstrap-3.2.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="Duke NGS Course (Summer 2015) 1.0 documentation" href="index.html" />
    <link rel="prev" title="Coding Exercises" href="PracticeONESolutios.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body role="document">

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          Duke NGS Course (Summer 2015)</a>
        <span class="navbar-text navbar-version pull-left"><b>1.0</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="VMsAtDuke.html">How to Claim and Access Your Virtual Machine</a></li>
<li class="toctree-l1"><a class="reference internal" href="Installation.html">Installing R, RStudio and IPython notebook with the R kernel</a></li>
<li class="toctree-l1"><a class="reference internal" href="BasicRinJupyterAndRstudio.html">Basic R in the Jupyter Notebook and RStudio</a></li>
<li class="toctree-l1"><a class="reference internal" href="IntroductionToR.html">Introduction to R</a></li>
<li class="toctree-l1"><a class="reference internal" href="PreparingDataForAnalysis.html">Preparing Data for Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="WorkingWithData.html">Working with Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="GroupingandAggregation.html">Grouping and Aggregation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Hypothesistestingandpowercalcuations.html">Hypothesis Testing and Power Calculations</a></li>
<li class="toctree-l1"><a class="reference internal" href="Probabilitydistributionsandrandomnumbergeneration.html">Probability distributions and Random Number Genereation</a></li>
<li class="toctree-l1"><a class="reference internal" href="WritingcustomfunctionsinRSolutions.html">Writing Custom Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="FunctionalProgramming.html">Functional Programming</a></li>
<li class="toctree-l1"><a class="reference internal" href="RFormulasAndStatisticalModeling.html">Linear Regression</a></li>
<li class="toctree-l1"><a class="reference internal" href="SupervisedLearning.html">Using R for supervised learning</a></li>
<li class="toctree-l1"><a class="reference internal" href="UnsupervisedLearning.html">Unsupervised Learning</a></li>
<li class="toctree-l1"><a class="reference internal" href="UnsupervisedLearningAndNGS.html">Unsupervised Learning and NGS</a></li>
<li class="toctree-l1"><a class="reference internal" href="MultipleTesting.html">Multiple Testing</a></li>
<li class="toctree-l1"><a class="reference internal" href="CountModels.html">Counting Models/Discrete Distributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="LogisticRegression.html">Generalized Linear Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="BaseGraphics.html">Base Graphics</a></li>
<li class="toctree-l1"><a class="reference internal" href="BaseGraphicsGGPlotComparison.html">Comparing Base Graphics with  <code class="docutils literal"><span class="pre">ggplot2</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="PracticeONESolutios.html">Coding Exercises</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Introduction to DESeq2</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Introduction to DESeq2</a><ul>
<li><a class="reference internal" href="#importing-and-inspecting-data">Importing and Inspecting Data</a></li>
<li><a class="reference internal" href="#estimate-size-factors-and-dispersion-parameters">Estimate Size Factors and Dispersion Parameters</a></li>
<li><a class="reference internal" href="#size-factors">Size Factors</a></li>
<li><a class="reference internal" href="#dispersion-parameters">Dispersion Parameters</a></li>
<li><a class="reference internal" href="#differential-expression-analysis">Differential Expression Analysis</a></li>
<li><a class="reference internal" href="#converting-normalizing-counts-to-expressions">Converting/Normalizing Counts to &#8220;Expressions&#8221;</a></li>
<li><a class="reference internal" href="#fpm">FPM</a></li>
<li><a class="reference internal" href="#fpkm">FPKM</a></li>
<li><a class="reference internal" href="#regularized-log-transformation">Regularized log transformation</a></li>
<li><a class="reference internal" href="#import-count-data-using-basic-tools">Import count data using basic tools</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
  <li>
    <a href="PracticeONESolutios.html" title="Previous Chapter: Coding Exercises"><span class="glyphicon glyphicon-chevron-left visible-sm"></span><span class="hidden-sm hidden-tablet">&laquo; Coding Exercises</span>
    </a>
  </li>
              
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="_sources/DESeq2Notebook.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="introduction-to-deseq2">
<h1>Introduction to DESeq2<a class="headerlink" href="#introduction-to-deseq2" title="Permalink to this headline">¶</a></h1>
<p>This notebook serves as a tutorial for using the DESeq2 package. Please
be sure to consult the excellent vignette provided by the DESeq2
package. Hopefully, we will also get a chance to review the edgeR
package (which also has a very nice vignette which I suggest that you
review)</p>
<p><strong>Load packages</strong></p>
<p>Load requisite R packages</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">library</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">tools</span><span class="p">)</span>
<span class="n">options</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in library(DESeq2): there is no package called ‘DESeq2’
</pre></div>
</div>
<div class="section" id="importing-and-inspecting-data">
<h2>Importing and Inspecting Data<a class="headerlink" href="#importing-and-inspecting-data" title="Permalink to this headline">¶</a></h2>
<p>Let&#8217;s set the file name containing the phenotype data and the directory
storing the count files from the htseq-count step. You will have to
adjust these strings. Note that it is assumed that all count files are
stored under a single directory.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">phfile</span><span class="o">=</span><span class="s">&quot;sampletable.txt&quot;</span>
<span class="n">cntdir</span><span class="o">=</span><span class="s">&quot;/home/owzar001/CURRENT/summercourse-2015/Data/COUNTS&quot;</span>
</pre></div>
</div>
<p>Next, read in the phenotype data from the sample file. It is always a
good idea to display the md5sum hash key for the file. Note that when
importing the file using read.table(), the stringsAsFactor argument is
set to FALSE. This imports strings as character rather than factor
objects.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">md5sum</span><span class="p">(</span><span class="n">phfile</span><span class="p">)</span>
<span class="n">phdata</span><span class="o">=</span><span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">phfile</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;,&quot;</span><span class="p">,</span><span class="n">stringsAsFactor</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
</pre></div>
</div>
<strong>sampletable.txt:</strong> '10cdaaa6e5fc76ab92b6d845d33a2a2f'<p>It is always a good idea to check the dimension of the file you have
read in</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">dim</span><span class="p">(</span><span class="n">phdata</span><span class="p">)</span>
</pre></div>
</div>
<ol class=list-inline>
    <li>6</li>
    <li>3</li>
</ol><p>Finally, it is a good idea to print out the sample data (or at least the
first few rows)</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">phdata</span>
</pre></div>
</div>
<table>
<thead><tr><th></th><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th></tr></thead>
<tbody>
    <tr><th scope=row>1</th><td>AGTCAA_counts.tsv</td><td>AGTCAA</td><td>0</td></tr>
    <tr><th scope=row>2</th><td>AGTTCC_counts.tsv</td><td>AGTTCC</td><td>0</td></tr>
    <tr><th scope=row>3</th><td>ATGTCA_counts.tsv</td><td>ATGTCA</td><td>0</td></tr>
    <tr><th scope=row>4</th><td>CCGTCC_counts.tsv</td><td>CCGTCC</td><td>1</td></tr>
    <tr><th scope=row>5</th><td>GTCCGC_counts.tsv</td><td>GTCCGC</td><td>1</td></tr>
    <tr><th scope=row>6</th><td>GTGAAA_counts.tsv</td><td>GTGAAA</td><td>1</td></tr>
</tbody>
</table><p>Next, we reformat the phenotype data object with more informative column
names and by adding the md5sum hash keys for the count files. When
dealing with directory and file names you should use file.path(),
dirname() and basename(). Check out the help files</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">colnames</span><span class="p">(</span><span class="n">phdata</span><span class="p">)</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s">&quot;filename&quot;</span><span class="p">,</span><span class="s">&quot;sampid&quot;</span><span class="p">,</span><span class="s">&quot;trt&quot;</span><span class="p">)</span>
<span class="n">phdata</span><span class="p">[[</span><span class="s">&quot;md5sum&quot;</span><span class="p">]]</span><span class="o">=</span><span class="n">md5sum</span><span class="p">(</span><span class="nb">file</span><span class="o">.</span><span class="n">path</span><span class="p">(</span><span class="n">cntdir</span><span class="p">,</span><span class="n">phdata</span><span class="p">[[</span><span class="s">&quot;filename&quot;</span><span class="p">]]))</span>
<span class="n">phdata</span>
</pre></div>
</div>
<table>
<thead><tr><th></th><th scope=col>filename</th><th scope=col>sampid</th><th scope=col>trt</th><th scope=col>md5sum</th></tr></thead>
<tbody>
    <tr><th scope=row>1</th><td>AGTCAA_counts.tsv</td><td>AGTCAA</td><td>0</td><td>NA</td></tr>
    <tr><th scope=row>2</th><td>AGTTCC_counts.tsv</td><td>AGTTCC</td><td>0</td><td>NA</td></tr>
    <tr><th scope=row>3</th><td>ATGTCA_counts.tsv</td><td>ATGTCA</td><td>0</td><td>NA</td></tr>
    <tr><th scope=row>4</th><td>CCGTCC_counts.tsv</td><td>CCGTCC</td><td>1</td><td>NA</td></tr>
    <tr><th scope=row>5</th><td>GTCCGC_counts.tsv</td><td>GTCCGC</td><td>1</td><td>NA</td></tr>
    <tr><th scope=row>6</th><td>GTGAAA_counts.tsv</td><td>GTGAAA</td><td>1</td><td>NA</td></tr>
</tbody>
</table><p>The first step to an analysis using the DESeq2 package is to import the
raw counts. If these counts stored in files generated by htseq-count,
then you may use the DESeqDataSetFromHTSeqCount() function from the
package. This function expects a sample table that contains the sample
id in the first column and the count file name in teh second column. Our
sample table is not in that format. It is easy to reorder the columns</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">phdata</span><span class="o">=</span><span class="n">phdata</span><span class="p">[</span><span class="n">c</span><span class="p">(</span><span class="s">&quot;sampid&quot;</span><span class="p">,</span><span class="s">&quot;filename&quot;</span><span class="p">,</span><span class="s">&quot;trt&quot;</span><span class="p">,</span><span class="s">&quot;md5sum&quot;</span><span class="p">)]</span>
<span class="n">phdata</span>
</pre></div>
</div>
<table>
<thead><tr><th></th><th scope=col>sampid</th><th scope=col>filename</th><th scope=col>trt</th><th scope=col>md5sum</th></tr></thead>
<tbody>
    <tr><th scope=row>1</th><td>AGTCAA</td><td>AGTCAA_counts.tsv</td><td>0</td><td>NA</td></tr>
    <tr><th scope=row>2</th><td>AGTTCC</td><td>AGTTCC_counts.tsv</td><td>0</td><td>NA</td></tr>
    <tr><th scope=row>3</th><td>ATGTCA</td><td>ATGTCA_counts.tsv</td><td>0</td><td>NA</td></tr>
    <tr><th scope=row>4</th><td>CCGTCC</td><td>CCGTCC_counts.tsv</td><td>1</td><td>NA</td></tr>
    <tr><th scope=row>5</th><td>GTCCGC</td><td>GTCCGC_counts.tsv</td><td>1</td><td>NA</td></tr>
    <tr><th scope=row>6</th><td>GTGAAA</td><td>GTGAAA_counts.tsv</td><td>1</td><td>NA</td></tr>
</tbody>
</table><p>Also, DESeq2 prefers that the treatment variable is of factor class. So
we will convert it</p>
<div class="code python highlight-python"><div class="highlight"><pre>phdata[[&quot;trt&quot;]]=as.factor(phdata[[&quot;trt&quot;]])
</pre></div>
</div>
<p>Now, we import the counts. Note that the first argument is the sample
table while the second is the directory storing the count files. The
last argument specifies the design. More on this later.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">dds</span><span class="o">=</span><span class="n">DESeqDataSetFromHTSeqCount</span><span class="p">(</span><span class="n">sampleTable</span> <span class="o">=</span> <span class="n">phdata</span><span class="p">,</span><span class="n">directory</span> <span class="o">=</span> <span class="n">cntdir</span><span class="p">,</span><span class="n">design</span><span class="o">=~</span><span class="n">trt</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;DESeqDataSetFromHTSeqCount&quot;
</pre></div>
</div>
<p><strong>Inspect object</strong></p>
<p>Let&#8217;s has a look at the object we have created.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">dds</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;dds&#39; not found
</pre></div>
</div>
<p>Note that this object is of class DESeqDataSet. It contains data on 4436
genes on 6 samples. Use the colData() function to see what you have read
in</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">colData</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;colData&quot;
</pre></div>
</div>
<p>The first thing you may want to do is to have a look at the raw counts
you have imported. You can use the counts(). Let&#8217;s looks the first three
genes (use the head() function to avoid printing out all genes). Before
that notw the dimension of the count matrix (does it look correct?)</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">dim</span><span class="p">(</span><span class="n">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;counts&quot;
</pre></div>
</div>
<p>Now print the raw counts for the first three genes (how can you verify
this looking at the files from htseq-count)</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">head</span><span class="p">(</span><span class="n">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">),</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in head(counts(dds), 3): could not find function &quot;counts&quot;
</pre></div>
</div>
<p><strong>8Slots of an S4 class</strong></p>
<p>To get the slots of an S4 class use slotNames()</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">slotNames</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in is(x, &quot;classRepresentation&quot;): object &#39;dds&#39; not found
</pre></div>
</div>
<p>Now let&#8217;s look at a few slots:</p>
<div class="code python highlight-python"><div class="highlight"><pre>This gives the design of the study
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in parse(text = x, srcfile = src): &lt;text&gt;:1:6: unexpected symbol
1: This gives
         ^
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre>dds@design
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;dds&#39; not found
</pre></div>
</div>
<p>The dispersion function is NULl for now (more on this later)</p>
<div class="code python highlight-python"><div class="highlight"><pre>dds@dispersionFunction
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;dds&#39; not found
</pre></div>
</div>
<p>This slots return gene specific information (it will be populated later)</p>
<div class="code python highlight-python"><div class="highlight"><pre>dds@rowData
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;dds&#39; not found
</pre></div>
</div>
<p>This slot returns the sample data</p>
<div class="code python highlight-python"><div class="highlight"><pre>dds@colData
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;dds&#39; not found
</pre></div>
</div>
<p>Before, going to the next step, let&#8217;s look at the output from mcols()</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">mcols</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;mcols&quot;
</pre></div>
</div>
</div>
<div class="section" id="estimate-size-factors-and-dispersion-parameters">
<h2>Estimate Size Factors and Dispersion Parameters<a class="headerlink" href="#estimate-size-factors-and-dispersion-parameters" title="Permalink to this headline">¶</a></h2>
<p>You recall that DESeq requires that we have estimates for sample
specific size factors and gene specific dispersion factors. More
specifically, recall that DESeq models the count <span class="math">\(K_{ij}\)</span> (gene
<span class="math">\(i\)</span>, sample <span class="math">\(j\)</span>) as negative binomial with mean
<span class="math">\(\mu_{ij}\)</span> and dispersion parameter <span class="math">\(\alpha_i\)</span>. Here
<span class="math">\(\mu_{ij}=s_j q_{ij}\)</span> where
<span class="math">\(\log_2(q_{ij}) = \beta_{0i} + \beta_{1i} z_j\)</span>. Here <span class="math">\(s_j\)</span>
is the sample <span class="math">\(j\)</span> specific size factor.</p>
</div>
<div class="section" id="size-factors">
<h2>Size Factors<a class="headerlink" href="#size-factors" title="Permalink to this headline">¶</a></h2>
<p>We begin by estimating the size factors <span class="math">\(s_1,\ldots,s_6\)</span>:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">dds</span> <span class="o">&lt;-</span> <span class="n">estimateSizeFactors</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;estimateSizeFactors&quot;
</pre></div>
</div>
<p>Now, compare the dds object to that of before applying the
estimateSizeFactors() function. What has changed? What remains
unchanged?</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">dds</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;dds&#39; not found
</pre></div>
</div>
<p>Note that there is a sizeFactor added to colData. Let&#8217;s look at it more
carefully</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">colData</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;colData&quot;
</pre></div>
</div>
<p>You can also get the size factors directly (why are there six size
factors?)</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">sizeFactors</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;sizeFactors&quot;
</pre></div>
</div>
<p>It is preferable to limit the number of decimal places. Next show the
size factors rounded to 3 decimal places</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="nb">round</span><span class="p">(</span><span class="n">sizeFactors</span><span class="p">(</span><span class="n">dds</span><span class="p">),</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;sizeFactors&quot;
</pre></div>
</div>
<p>Now that the size factors have been estimated, we can get &#8220;normalized&#8221;
counts</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">head</span><span class="p">(</span><span class="n">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">),</span><span class="mi">3</span><span class="p">)</span>
<span class="n">head</span><span class="p">(</span><span class="n">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="n">normalize</span><span class="o">=</span><span class="n">TRUE</span><span class="p">),</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in head(counts(dds), 3): could not find function &quot;counts&quot;
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in head(counts(dds, normalize = TRUE), 3): could not find function &quot;counts&quot;
</pre></div>
</div>
<p>Note that these are the counts divided by the size factors. Compare the
first row of the last table (&#8220;normalized&#8221; counts for gene 1) to the hand
calculation below.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">)[</span><span class="mi">1</span><span class="p">,]</span><span class="o">/</span><span class="n">sizeFactors</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;counts&quot;
</pre></div>
</div>
<p>Exercise: How do you get the raw counts for gene &#8220;GeneID:12930116&#8221;?</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">)[</span><span class="s">&quot;GeneID:12930116&quot;</span><span class="p">,]</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;counts&quot;
</pre></div>
</div>
<p>Exercise: How do you get the normalized counts for gene
&#8220;GeneID:12930116&#8221;?</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="n">normalize</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)[</span><span class="s">&quot;GeneID:12930116&quot;</span><span class="p">,]</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;counts&quot;
</pre></div>
</div>
<p>Exercise: Get a summary (mean, median, quantiles etc ) of the size
factors</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">summary</span><span class="p">(</span><span class="n">sizeFactors</span><span class="p">(</span><span class="n">dds</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in summary(sizeFactors(dds)): could not find function &quot;sizeFactors&quot;
</pre></div>
</div>
<p>Before going to the next step, let&#8217;s look at the dispersionFunction slot</p>
<div class="code python highlight-python"><div class="highlight"><pre>dds@dispersionFunction
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;dds&#39; not found
</pre></div>
</div>
</div>
<div class="section" id="dispersion-parameters">
<h2>Dispersion Parameters<a class="headerlink" href="#dispersion-parameters" title="Permalink to this headline">¶</a></h2>
<p>Next, we get the dispersion factors
<span class="math">\(\alpha_1,\ldots,\alpha_{4436}\)</span></p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">dds</span><span class="o">=</span><span class="n">estimateDispersions</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;estimateDispersions&quot;
</pre></div>
</div>
<p>Now inspect the dds object again and note that the rowRanges slot has
extra information (&#8220;metadata column names(0):&#8221; before versus &#8220;column
names(9): baseMean baseVar ... dispOutlier dispMAP&#8221;)</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">dds</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;dds&#39; not found
</pre></div>
</div>
<p>We can extract the gene specific dispersion factors using dispersions().
Note that there will be one number per gene. We look at the first four
genes (rounded to 4 decimal places)</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">alphas</span><span class="o">=</span><span class="n">dispersions</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;dispersions&quot;
</pre></div>
</div>
<p>Verify that the number of dispersion factors equals the number of genes</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">length</span><span class="p">(</span><span class="n">alphas</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;alphas&#39; not found
</pre></div>
</div>
<p>Print the dispersion factors for the first 5 genes rounded to four
decimal points</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="nb">round</span><span class="p">(</span><span class="n">alphas</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">5</span><span class="p">],</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;alphas&#39; not found
</pre></div>
</div>
<p>Extract the metadata using mcols() for the first four genes (recall that
it was previously</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">mcols</span><span class="p">(</span><span class="n">dds</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="mi">4</span><span class="p">,]</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;mcols&quot;
</pre></div>
</div>
<p>Exercise: Provide statistical summaries of the dispersion factors</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">summary</span><span class="p">(</span><span class="n">dispersions</span><span class="p">(</span><span class="n">dds</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in summary(dispersions(dds)): could not find function &quot;dispersions&quot;
</pre></div>
</div>
<p>Exercise: Summarize the dispersion factors using a box plot (may want to
log transform)</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">boxplot</span><span class="p">(</span><span class="n">log</span><span class="p">(</span><span class="n">dispersions</span><span class="p">(</span><span class="n">dds</span><span class="p">)))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/DESeq2Notebook_86_0.png"><img alt="_images/DESeq2Notebook_86_0.png" src="_images/DESeq2Notebook_86_0.png" style="width: 600px;" /></a>
<div class="highlight-python"><div class="highlight"><pre>Error in boxplot(log(dispersions(dds))): could not find function &quot;dispersions&quot;
</pre></div>
</div>
</div>
<div class="section" id="differential-expression-analysis">
<h2>Differential Expression Analysis<a class="headerlink" href="#differential-expression-analysis" title="Permalink to this headline">¶</a></h2>
<p>We can now conduct a differential expression analysis using the DESeq()
function. Keep in mind that to get to this step, we first estimated the
size factors and then the dispersion parameters.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">ddsDE</span><span class="o">=</span><span class="n">DESeq</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;DESeq&quot;
</pre></div>
</div>
<p>We can get the results for the differential expression analysis using
results()</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">myres</span><span class="o">=</span><span class="n">results</span><span class="p">(</span><span class="n">ddsDE</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;results&quot;
</pre></div>
</div>
<p>Let&#8217;s look at the results for the first four genes</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">myres</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">4</span><span class="p">,]</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;myres&#39; not found
</pre></div>
</div>
<p>Calculate BH adjusted P-values by &#8220;hand&#8221; using the p.adjust() function</p>
<div class="code python highlight-python"><div class="highlight"><pre>BH=p.adjust(myres$pvalue,&quot;BH&quot;)
BH[1:4]
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in p.adjust(myres$pvalue, &quot;BH&quot;): object &#39;myres&#39; not found
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;BH&#39; not found
</pre></div>
</div>
<p>You can get the descriptions for the columns from the DE analysis</p>
<div class="code python highlight-python"><div class="highlight"><pre>data.frame(desc=mcols(myres)$description)
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in data.frame(desc = mcols(myres)$description): could not find function &quot;mcols&quot;
</pre></div>
</div>
<p>We can get summaries of the results:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">summary</span><span class="p">(</span><span class="n">myres</span><span class="p">,</span><span class="mf">0.05</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in summary(myres, 0.05): object &#39;myres&#39; not found
</pre></div>
</div>
<p>You can sort the results by say the unadjusted P-values</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">myres</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">myres</span><span class="p">[[</span><span class="s">&quot;pvalue&quot;</span><span class="p">]])[</span><span class="mi">1</span><span class="p">:</span><span class="mi">4</span><span class="p">],]</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;myres&#39; not found
</pre></div>
</div>
<p>To get the list of genes with unadjusted P-values &lt; 0.00001 and absolute
log2 FC of more than 4</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">subset</span><span class="p">(</span><span class="n">myres</span><span class="p">,</span><span class="n">pvalue</span><span class="o">&lt;</span><span class="mf">0.00001</span><span class="o">&amp;</span><span class="nb">abs</span><span class="p">(</span><span class="n">log2FoldChange</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in subset(myres, pvalue &lt; 1e-05 &amp; abs(log2FoldChange) &gt; 4): object &#39;myres&#39; not found
</pre></div>
</div>
<p>To get the list of genes with unadjusted P-values &lt; 0.00001 and
upregulated genes with log2 FC of more than 4</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">subset</span><span class="p">(</span><span class="n">myres</span><span class="p">,</span><span class="n">pvalue</span><span class="o">&lt;</span><span class="mf">0.00001</span><span class="o">&amp;</span><span class="n">log2FoldChange</span><span class="o">&gt;</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in subset(myres, pvalue &lt; 1e-05 &amp; log2FoldChange &gt; 4): object &#39;myres&#39; not found
</pre></div>
</div>
<p>The P-values for the four top genes are beyond machine precision. You
can use the format.pval() function to properly format the P-values.
PLEASE promote ending the practice of publishing P-values below machine
precision. (that would be akin to stating the weight of an object that
weighs less than one pound with scale that whose minimum weight spec is
1lbs).</p>
<div class="code python highlight-python"><div class="highlight"><pre>myres$pval=format.pval(myres$pvalue)
myres[order(myres[[&quot;pvalue&quot;]])[1:4],]
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in format.pval(myres$pvalue): object &#39;myres&#39; not found
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;myres&#39; not found
</pre></div>
</div>
<p>Let&#8217;s look at a volcano plot</p>
<div class="code python highlight-python"><div class="highlight"><pre>plot(myres$log2FoldChange,-log10(myres$padj),pch=19,cex=0.3,xlab=&quot;Log2 FC&quot;,ylab=&quot;-log10(BH Adjusted P-value)&quot;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/DESeq2Notebook_109_0.png"><img alt="_images/DESeq2Notebook_109_0.png" src="_images/DESeq2Notebook_109_0.png" style="width: 600px;" /></a>
<div class="highlight-python"><div class="highlight"><pre>Error in plot(myres$log2FoldChange, -log10(myres$padj), pch = 19, cex = 0.3, : object &#39;myres&#39; not found
</pre></div>
</div>
<p>Extract results for genes GeneID:12932226 and GeneID:12930116</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">myres</span><span class="p">[</span><span class="n">c</span><span class="p">(</span><span class="s">&quot;GeneID:12932226&quot;</span><span class="p">,</span><span class="s">&quot;GeneID:12930116&quot;</span><span class="p">),]</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;myres&#39; not found
</pre></div>
</div>
<p>Exercise: Annotate the hits with adjusted P-values &lt; 0.05 and absolute
log2 FC greater than 2 in red</p>
<div class="code python highlight-python"><div class="highlight"><pre>plot(myres$log2FoldChange,-log10(myres$padj),pch=19,cex=0.3,xlab=&quot;Log2 FC&quot;,ylab=&quot;-log10(BH Adjusted P-value)&quot;,col=ifelse(myres$padj&lt;0.05&amp;abs(myres$log2FoldChange)&gt;2,&quot;red&quot;,&quot;black&quot;))
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/DESeq2Notebook_113_0.png"><img alt="_images/DESeq2Notebook_113_0.png" src="_images/DESeq2Notebook_113_0.png" style="width: 600px;" /></a>
<div class="highlight-python"><div class="highlight"><pre>Error in plot(myres$log2FoldChange, -log10(myres$padj), pch = 19, cex = 0.3, : object &#39;myres&#39; not found
</pre></div>
</div>
</div>
<div class="section" id="converting-normalizing-counts-to-expressions">
<h2>Converting/Normalizing Counts to &#8220;Expressions&#8221;<a class="headerlink" href="#converting-normalizing-counts-to-expressions" title="Permalink to this headline">¶</a></h2>
<p><strong>Normalized Counts</strong></p>
<p>We have already shown how to &#8220;normalize&#8221; the counts using the estimated
size factors</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">head</span><span class="p">(</span><span class="n">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="n">normalize</span><span class="o">=</span><span class="n">TRUE</span><span class="p">),</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in head(counts(dds, normalize = TRUE), 3): could not find function &quot;counts&quot;
</pre></div>
</div>
<p>Plot the counts stratified by treatment for the 2nd gene</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plotCounts</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span><span class="n">intgroup</span><span class="o">=</span><span class="s">&quot;trt&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/DESeq2Notebook_119_0.png"><img alt="_images/DESeq2Notebook_119_0.png" src="_images/DESeq2Notebook_119_0.png" style="width: 600px;" /></a>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;plotCounts&quot;
</pre></div>
</div>
<p>Or alternatively (better)</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plotCounts</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span> <span class="s">&quot;GeneID:12930115&quot;</span><span class="p">,</span><span class="n">intgroup</span><span class="o">=</span><span class="s">&quot;trt&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/DESeq2Notebook_121_0.png"><img alt="_images/DESeq2Notebook_121_0.png" src="_images/DESeq2Notebook_121_0.png" style="width: 600px;" /></a>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;plotCounts&quot;
</pre></div>
</div>
<p>Now get this plot for the top hit</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plotCounts</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span> <span class="s">&quot;GeneID:12931678&quot;</span><span class="p">,</span><span class="n">intgroup</span><span class="o">=</span><span class="s">&quot;trt&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/DESeq2Notebook_123_0.png"><img alt="_images/DESeq2Notebook_123_0.png" src="_images/DESeq2Notebook_123_0.png" style="width: 600px;" /></a>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;plotCounts&quot;
</pre></div>
</div>
</div>
<div class="section" id="fpm">
<h2>FPM<a class="headerlink" href="#fpm" title="Permalink to this headline">¶</a></h2>
<div class="code python highlight-python"><div class="highlight"><pre>Another approach is to FPM: fragments per million mapped fragments
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in parse(text = x, srcfile = src): &lt;text&gt;:1:9: unexpected symbol
1: Another approach
            ^
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">head</span><span class="p">(</span><span class="n">fpm</span><span class="p">(</span><span class="n">dds</span><span class="p">),</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in head(fpm(dds), 3): could not find function &quot;fpm&quot;
</pre></div>
</div>
<p>Let&#8217;s calculate the FPM manually. For gene <span class="math">\(i\)</span> sample <span class="math">\(j\)</span>,
the FPM is defined as <span class="math">\(\frac{K_{ij}}{D_j}\times 10^{6}\)</span> where
<span class="math">\(D_j=\sum_{i=1} K_{ij}\)</span> is the read depth for sample <span class="math">\(j\)</span>.
First get the read depth for each sample</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">D</span><span class="o">=</span><span class="n">colSums</span><span class="p">(</span><span class="n">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">))</span>
<span class="n">D</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in is.data.frame(x): could not find function &quot;counts&quot;
</pre></div>
</div>
<pre class=language-r><code>function (expr, name)
.External(C_doD, expr, name)</code></pre><p>By default, the fpm() function uses a robust approach. We will disable
this right now as to replicate the standard FPM. Let&#8217;s look at gene 1</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">fpm1</span><span class="o">=</span><span class="n">fpm</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="n">robust</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)[</span><span class="mi">1</span><span class="p">,]</span>
<span class="n">fpm1</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;fpm&quot;
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;fpm1&#39; not found
</pre></div>
</div>
<p>Now get the raw counts for gene 1</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">cnt1</span><span class="o">=</span><span class="n">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">)[</span><span class="mi">1</span><span class="p">,]</span>
<span class="n">cnt1</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;counts&quot;
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;cnt1&#39; not found
</pre></div>
</div>
<p>Now calculate the FPM for gene 1</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">myfpm1</span><span class="o">=</span><span class="n">cnt1</span><span class="o">/</span><span class="n">D</span><span class="o">*</span><span class="mf">1e6</span>
<span class="n">myfpm1</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;cnt1&#39; not found
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;myfpm1&#39; not found
</pre></div>
</div>
<p>This is how you check if two numeric columns are &#8220;equal&#8221;</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="nb">min</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">fpm1</span><span class="o">-</span><span class="n">myfpm1</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): object &#39;fpm1&#39; not found
</pre></div>
</div>
</div>
<div class="section" id="fpkm">
<h2>FPKM<a class="headerlink" href="#fpkm" title="Permalink to this headline">¶</a></h2>
<p>To calculate the FPKM (fragments per kilobase per million mapped
fragments) we need to add annotation to assign the feature lengths. More
specifically, for gene <span class="math">\(i\)</span> sample <span class="math">\(j\)</span>, the FPKM is defined
as <span class="math">\(\frac{K_{ij}}{\ell_i D_j}\times 10^3 \times 10^{6}\)</span> where
<span class="math">\(\ell_i\)</span> is the &#8220;length&#8221; of gene <span class="math">\(i\)</span> (fragments for each
<span class="math">\(10^3\)</span> bases in the gene for every <span class="math">\(\frac{D_j}{10^6}\)</span>
fragments. More on this later.</p>
</div>
<div class="section" id="regularized-log-transformation">
<h2>Regularized log transformation<a class="headerlink" href="#regularized-log-transformation" title="Permalink to this headline">¶</a></h2>
<p>The regularized log transform can be obtained using the rlog() function.
Note that an important argument for this function is blind (TRUE by
default). The default &#8220;blinds&#8221; the normalization to the design. This is
very important so as to not bias the analyses (e.g. class discovery</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">rld</span><span class="o">=</span><span class="n">rlog</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="n">blind</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;rlog&quot;
</pre></div>
</div>
<p>Hierarchical clustering using rlog transformation</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">dists</span><span class="o">=</span><span class="n">dist</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">assay</span><span class="p">(</span><span class="n">rld</span><span class="p">)))</span>
<span class="n">plot</span><span class="p">(</span><span class="n">hclust</span><span class="p">(</span><span class="n">dists</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/DESeq2Notebook_143_0.png"><img alt="_images/DESeq2Notebook_143_0.png" src="_images/DESeq2Notebook_143_0.png" style="width: 600px;" /></a>
<div class="highlight-python"><div class="highlight"><pre>Error in t(assay(rld)): could not find function &quot;assay&quot;
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Error in hclust(dists): object &#39;dists&#39; not found
</pre></div>
</div>
<p>PC Analysis using the rlog transformation</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plotPCA</span><span class="p">(</span><span class="n">rld</span><span class="p">,</span><span class="n">intgroup</span><span class="o">=</span><span class="s">&quot;trt&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/DESeq2Notebook_145_0.png"><img alt="_images/DESeq2Notebook_145_0.png" src="_images/DESeq2Notebook_145_0.png" style="width: 600px;" /></a>
<div class="highlight-python"><div class="highlight"><pre>Error in eval(expr, envir, enclos): could not find function &quot;plotPCA&quot;
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">sessionInfo</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>R version 3.1.2 (2014-10-31)
Platform: x86_64-apple-darwin13.4.0 (64-bit)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] tools     stats     graphics  grDevices utils     datasets  methods   base

loaded via a namespace (and not attached):
 [1] base64enc_0.1-3      digest_0.6.8         evaluate_0.7.2       IRdisplay_0.3.0.9000
 [5] IRkernel_0.3.0.9000  jsonlite_0.9.17      magrittr_1.5         repr_0.1.0.9000
 [9] rzmq_0.7.7           stringi_0.5-5        stringr_1.0.0        uuid_0.1-2
</pre></div>
</div>
</div>
<div class="section" id="import-count-data-using-basic-tools">
<h2>Import count data using basic tools<a class="headerlink" href="#import-count-data-using-basic-tools" title="Permalink to this headline">¶</a></h2>
<p>We want to read in the count files using more basic tools (important if
the count files are not generated fro,m htseq-count). We will import the
files as matrices. We start with some of the preliminary steps</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">library</span><span class="p">(</span><span class="n">tools</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">phfile</span><span class="o">=</span><span class="s">&quot;/home/owzar001/CURRENT/summercourse-2015/Data/sampletable.txt&quot;</span>
<span class="n">cntdir</span><span class="o">=</span><span class="s">&quot;/home/owzar001/CURRENT/summercourse-2015/Data/COUNTS&quot;</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">md5sum</span><span class="p">(</span><span class="n">phfile</span><span class="p">)</span>
<span class="n">phdata</span><span class="o">=</span><span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">phfile</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;,&quot;</span><span class="p">,</span><span class="n">stringsAsFactor</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
</pre></div>
</div>
<strong>/home/owzar001/CURRENT/summercourse-2015/Data/sampletable.txt:</strong> '10cdaaa6e5fc76ab92b6d845d33a2a2f'<div class="code python highlight-python"><div class="highlight"><pre><span class="n">colnames</span><span class="p">(</span><span class="n">phdata</span><span class="p">)</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s">&quot;filename&quot;</span><span class="p">,</span><span class="s">&quot;sampid&quot;</span><span class="p">,</span><span class="s">&quot;trt&quot;</span><span class="p">)</span>
<span class="n">phdata</span><span class="p">[[</span><span class="s">&quot;md5sum&quot;</span><span class="p">]]</span><span class="o">=</span><span class="n">md5sum</span><span class="p">(</span><span class="nb">file</span><span class="o">.</span><span class="n">path</span><span class="p">(</span><span class="n">cntdir</span><span class="p">,</span><span class="n">phdata</span><span class="p">[[</span><span class="s">&quot;filename&quot;</span><span class="p">]]))</span>
<span class="n">phdata</span>
</pre></div>
</div>
<table>
<thead><tr><th></th><th scope=col>filename</th><th scope=col>sampid</th><th scope=col>trt</th><th scope=col>md5sum</th></tr></thead>
<tbody>
    <tr><th scope=row>1</th><td>AGTCAA_counts.tsv</td><td>AGTCAA</td><td>0</td><td>a9eaa959aba1b02b3831583c2a9751c8</td></tr>
    <tr><th scope=row>2</th><td>AGTTCC_counts.tsv</td><td>AGTTCC</td><td>0</td><td>4183767e4eeb75dc582bcf438af13500</td></tr>
    <tr><th scope=row>3</th><td>ATGTCA_counts.tsv</td><td>ATGTCA</td><td>0</td><td>26fbba06520758e5a3acd9bd432ebed4</td></tr>
    <tr><th scope=row>4</th><td>CCGTCC_counts.tsv</td><td>CCGTCC</td><td>1</td><td>50036a88fd48645f740a31f4f4352cfb</td></tr>
    <tr><th scope=row>5</th><td>GTCCGC_counts.tsv</td><td>GTCCGC</td><td>1</td><td>bb1cecd886127159157e9431d072cad5</td></tr>
    <tr><th scope=row>6</th><td>GTGAAA_counts.tsv</td><td>GTGAAA</td><td>1</td><td>fa544c0a076eedb54937c7189f4e1fbc</td></tr>
</tbody>
</table><div class="code python highlight-python"><div class="highlight"><pre><span class="n">phdata</span><span class="o">=</span><span class="n">phdata</span><span class="p">[</span><span class="n">c</span><span class="p">(</span><span class="s">&quot;sampid&quot;</span><span class="p">,</span><span class="s">&quot;filename&quot;</span><span class="p">,</span><span class="s">&quot;trt&quot;</span><span class="p">,</span><span class="s">&quot;md5sum&quot;</span><span class="p">)]</span>
<span class="n">phdata</span>
</pre></div>
</div>
<table>
<thead><tr><th></th><th scope=col>sampid</th><th scope=col>filename</th><th scope=col>trt</th><th scope=col>md5sum</th></tr></thead>
<tbody>
    <tr><th scope=row>1</th><td>AGTCAA</td><td>AGTCAA_counts.tsv</td><td>0</td><td>a9eaa959aba1b02b3831583c2a9751c8</td></tr>
    <tr><th scope=row>2</th><td>AGTTCC</td><td>AGTTCC_counts.tsv</td><td>0</td><td>4183767e4eeb75dc582bcf438af13500</td></tr>
    <tr><th scope=row>3</th><td>ATGTCA</td><td>ATGTCA_counts.tsv</td><td>0</td><td>26fbba06520758e5a3acd9bd432ebed4</td></tr>
    <tr><th scope=row>4</th><td>CCGTCC</td><td>CCGTCC_counts.tsv</td><td>1</td><td>50036a88fd48645f740a31f4f4352cfb</td></tr>
    <tr><th scope=row>5</th><td>GTCCGC</td><td>GTCCGC_counts.tsv</td><td>1</td><td>bb1cecd886127159157e9431d072cad5</td></tr>
    <tr><th scope=row>6</th><td>GTGAAA</td><td>GTGAAA_counts.tsv</td><td>1</td><td>fa544c0a076eedb54937c7189f4e1fbc</td></tr>
</tbody>
</table><div class="code python highlight-python"><div class="highlight"><pre>phdata[[&quot;trt&quot;]]=as.factor(phdata[[&quot;trt&quot;]])
</pre></div>
</div>
<p>This function reads the data from one file and returns the results as a
matrix. It is designed to only keep the rows for which the gene id
starts with the prefix &#8220;GeneID&#8221;</p>
<div class="code python highlight-python"><div class="highlight"><pre>readcnts=function(fname,sep=&quot;\t&quot;,prefix=&quot;GeneID&quot;,collab=&quot;V1&quot;,header=FALSE)
    {
        ### Import text file
        dat=read.table(fname,sep=sep,header=header,stringsAsFactor=FALSE)
        ### Only keep rows for which the gene id matches with the prefix
        dat=dat[substr(dat[[collab]],1,nchar(prefix))==prefix,]
        return(dat)
    }
</pre></div>
</div>
<p>Let&#8217;s read in a file</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">file1</span><span class="o">=</span><span class="n">readcnts</span><span class="p">(</span><span class="s">&quot;/home/owzar001/CURRENT/summercourse-2015/Data/COUNTS/AGTCAA_counts.tsv&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">dim</span><span class="p">(</span><span class="n">file1</span><span class="p">)</span>
</pre></div>
</div>
<ol class=list-inline>
    <li>4436</li>
    <li>2</li>
</ol><div class="code python highlight-python"><div class="highlight"><pre><span class="n">head</span><span class="p">(</span><span class="n">file1</span><span class="p">)</span>
</pre></div>
</div>
<table>
<thead><tr><th></th><th scope=col>V1</th><th scope=col>V2</th></tr></thead>
<tbody>
    <tr><th scope=row>1</th><td>GeneID:12930114</td><td>118</td></tr>
    <tr><th scope=row>2</th><td>GeneID:12930115</td><td>30</td></tr>
    <tr><th scope=row>3</th><td>GeneID:12930116</td><td>15</td></tr>
    <tr><th scope=row>4</th><td>GeneID:12930117</td><td>12</td></tr>
    <tr><th scope=row>5</th><td>GeneID:12930118</td><td>122</td></tr>
    <tr><th scope=row>6</th><td>GeneID:12930119</td><td>60</td></tr>
</tbody>
</table><p>Now read in all files</p>
<div class="code python highlight-python"><div class="highlight"><pre>## Read in the first file
countdat=readcnts(file.path(cntdir,phdata$filename[1]))
### Assign the sample id as the column name
names(countdat)[2]=phdata$sampid[1]
### Repeat the last two steps for files 2 ... 6 and merge
### along each step
for(i in 2:nrow(phdata))
    {
        dat2=readcnts(file.path(cntdir,phdata$filename[i]))
        names(dat2)[2]=phdata$sampid[i]
        countdat=merge(countdat,dat2,by=&quot;V1&quot;)
    }

geneid=countdat$V1
countdat=as.matrix(countdat[,-1])
row.names(countdat)=geneid
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">head</span><span class="p">(</span><span class="n">countdat</span><span class="p">)</span>
</pre></div>
</div>
<table>
<thead><tr><th></th><th scope=col>AGTCAA</th><th scope=col>AGTTCC</th><th scope=col>ATGTCA</th><th scope=col>CCGTCC</th><th scope=col>GTCCGC</th><th scope=col>GTGAAA</th></tr></thead>
<tbody>
    <tr><th scope=row>GeneID:12930114</th><td>118</td><td>137</td><td>149</td><td>120</td><td>161</td><td>174</td></tr>
    <tr><th scope=row>GeneID:12930115</th><td>30</td><td>42</td><td>25</td><td>18</td><td>32</td><td>34</td></tr>
    <tr><th scope=row>GeneID:12930116</th><td>15</td><td>55</td><td>37</td><td>49</td><td>36</td><td>27</td></tr>
    <tr><th scope=row>GeneID:12930117</th><td>12</td><td>12</td><td>13</td><td>11</td><td> 7</td><td> 6</td></tr>
    <tr><th scope=row>GeneID:12930118</th><td>122</td><td>137</td><td> 94</td><td> 48</td><td>131</td><td> 69</td></tr>
    <tr><th scope=row>GeneID:12930119</th><td>60</td><td>88</td><td>78</td><td>53</td><td>43</td><td>29</td></tr>
</tbody>
</table><div class="code python highlight-python"><div class="highlight"><pre><span class="n">library</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Loading required package: S4Vectors
Loading required package: stats4
Loading required package: BiocGenerics
Loading required package: parallel

Attaching package: ‘BiocGenerics’

The following objects are masked from ‘package:parallel’:

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB

The following object is masked from ‘package:stats’:

    xtabs

The following objects are masked from ‘package:base’:

    anyDuplicated, append, as.data.frame, as.vector, cbind, colnames,
    do.call, duplicated, eval, evalq, Filter, Find, get, intersect,
    is.unsorted, lapply, Map, mapply, match, mget, order, paste, pmax,
    pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce, rep.int,
    rownames, sapply, setdiff, sort, table, tapply, union, unique,
    unlist, unsplit

Creating a generic function for ‘nchar’ from package ‘base’ in package ‘S4Vectors’
Loading required package: IRanges
Loading required package: GenomicRanges
Loading required package: GenomeInfoDb
Loading required package: Rcpp
Loading required package: RcppArmadillo
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">dds</span><span class="o">=</span><span class="n">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">countdat</span><span class="p">,</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">phdata</span><span class="p">),</span><span class="n">design</span><span class="o">=~</span><span class="n">trt</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">dds</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>class: DESeqDataSet
dim: 4436 6
exptData(0):
assays(1): counts
rownames(4436): GeneID:12930114 GeneID:12930115 ... GeneID:13406005
  GeneID:13406006
rowRanges metadata column names(0):
colnames(6): AGTCAA AGTTCC ... GTCCGC GTGAAA
colData names(4): sampid filename trt md5sum
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">sessionInfo</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>R version 3.2.1 (2015-06-18)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 8 (jessie)

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
 [1] parallel  stats4    tools     stats     graphics  grDevices utils
 [8] datasets  methods   base

other attached packages:
[1] DESeq2_1.8.1              RcppArmadillo_0.5.200.1.0
[3] Rcpp_0.11.6               GenomicRanges_1.20.5
[5] GenomeInfoDb_1.4.1        IRanges_2.2.5
[7] S4Vectors_0.6.2           BiocGenerics_0.14.0

loaded via a namespace (and not attached):
 [1] RColorBrewer_1.1-2   futile.logger_1.4.1  plyr_1.8.3
 [4] XVector_0.8.0        futile.options_1.0.0 base64enc_0.1-3
 [7] rpart_4.1-9          digest_0.6.8         uuid_0.1-2
[10] RSQLite_1.0.0        annotate_1.46.1      lattice_0.20-31
[13] jsonlite_0.9.16      evaluate_0.7         gtable_0.1.2
[16] DBI_0.3.1            IRdisplay_0.3        IRkernel_0.4
[19] proto_0.3-10         gridExtra_2.0.0      rzmq_0.7.7
[22] genefilter_1.50.0    cluster_2.0.1        repr_0.3
[25] stringr_1.0.0        locfit_1.5-9.1       nnet_7.3-9
[28] grid_3.2.1           Biobase_2.28.0       AnnotationDbi_1.30.1
[31] XML_3.98-1.3         survival_2.38-2      BiocParallel_1.2.9
[34] foreign_0.8-63       latticeExtra_0.6-26  Formula_1.2-1
[37] geneplotter_1.46.0   ggplot2_1.0.1        reshape2_1.4.1
[40] lambda.r_1.1.7       magrittr_1.5         splines_3.2.1
[43] scales_0.2.5         Hmisc_3.16-0         MASS_7.3-40
[46] xtable_1.7-4         colorspace_1.2-6     stringi_0.5-5
[49] acepack_1.3-3.3      munsell_0.4.2
</pre></div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright Public Domain.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.3.1.<br/>
    </p>
  </div>
</footer>
  </body>
</html>